File: supabase/functions/agent-worker/agents/cross_source_reconciler.ts
// Cross-source reconciler agent.
//
// Catches the class of bug Ross caught manually on 2026-05-05: the
// dashboard's Net Sales KPI was inflated +20-30% over Toast's official
// "Net sales" report because the ETL was sourcing from `check.amount`
// (which empirically includes tips, service charges, etc.) instead of
// computing Net Sales from selection-level data.
//
// Two reconciliation checks per outlet, last 30 days:
//
//   1. INTERNAL: order_details.amount / order_details.net_sales ratio.
//      Expected ratio depends on outlet (driven by tip + service-charge
//      mix), but a healthy LSBR-style outlet sits around 1.20. We alert
//      if the ratio exits a sanity band [0.90, 1.50], which signals
//      either the new net_sales math broke OR check.amount started
//      including something we don't subtract (new Toast SC type, etc.).
//
//   2. EXTERNAL: order_details.net_sales vs sales_summary.net_sales on
//      the same date. These two are computed from different Toast data
//      paths (Orders API selections vs Sales Summary report import).
//      A persistent drift > 5% means one of the sources got stale or
//      misaligned — exactly the failure mode that motivated this agent.
//
// The internal check is dormant until net_sales is populated on enough
// daily rows (post-backfill). The external check only fires for outlets
// that have sales_summary baseline data (currently LSBR; will expand
// once the Sales Summary auto-pull lands).
//
// Audits: one entry per outlet × check pair (whether normal or alerting).
// Alerts: one event per breached threshold, routed via alert_dispatcher
// to the same Slack channel as drift/anomaly.

import type { AuditDecision } from "../lib/types.ts";

const PAGES_BASE = "https://rrmethodco.github.io/method-q1-2026-dashboards/data";
const OUTLETS = ["lsbr", "mulherins", "kampers", "lowland", "vessel",
  "anthology", "rosemary_rose", "hiroki_det", "hiroki_phl", "little_wing", "quoin"];

// Sanity bounds for amount/net_sales ratio. Below 0.90 means net_sales is
// somehow LARGER than amount (impossible if the formulas are right —
// indicates a sign error or schema regression). Above 1.50 means the
// gap between the two has blown out beyond a reasonable tip+SC overlay
// (typical LSBR ratio is 1.20-1.30 historically).
const RATIO_LOWER = 0.90;
const RATIO_UPPER = 1.50;

// Cross-source drift threshold: 5% sustained delta across the lookback
// window between two independently-computed Net Sales values means
// something is misaligned. Tighter than the in-period anomaly threshold
// because we're comparing aggregates over 30 days.
const CROSS_SOURCE_DRIFT = 0.05;

// Minimum coverage before the internal check is meaningful. Pre-backfill
// rows lack net_sales entirely — we wait until at least 7 days of the
// 30-day window have populated values before comparing.
const MIN_NET_SALES_DAYS = 7;

const LOOKBACK_DAYS = 30;

export interface ReconcilerAlert {
  outlet: string;
  kind: "internal_ratio_breach" | "external_source_drift" | "stale_baseline";
  text: string;
}

export interface ReconcilerResult {
  audits: AuditDecision[];
  alerts: ReconcilerAlert[];
}

interface DailyTotals {
  date: string;
  amount: number;
  net_sales: number;
  has_net_sales: boolean;
}

function lastNDays(n: number): string[] {
  const out: string[] = [];
  const today = new Date();
  for (let i = 1; i <= n; i++) {
    const d = new Date(today);
    d.setUTCDate(d.getUTCDate() - i);
    out.push(d.toISOString().slice(0, 10));
  }
  return out;
}

// Sum daily across all RC keys. Returns one row per date with both the
// legacy `amount` and Toast-aligned `net_sales`.
// deno-lint-ignore no-explicit-any
function aggregateOrderDetailsDaily(payload: any, dateRange: Set<string>): DailyTotals[] {
  const orderDetails = payload?.order_details;
  if (!orderDetails || typeof orderDetails !== "object") return [];
  const byDate: Record<string, DailyTotals> = {};
  for (const rcKey of Object.keys(orderDetails)) {
    const rcDaily = orderDetails[rcKey]?.daily;
    if (!Array.isArray(rcDaily)) continue;
    for (const row of rcDaily) {
      if (!row || typeof row.date !== "string") continue;
      if (!dateRange.has(row.date)) continue;
      const slot = byDate[row.date] ?? {
        date: row.date, amount: 0, net_sales: 0, has_net_sales: false,
      };
      const amt = Number(row.amount);
      const ns = row.net_sales != null ? Number(row.net_sales) : null;
      if (Number.isFinite(amt)) slot.amount += amt;
      if (ns != null && Number.isFinite(ns)) {
        slot.net_sales += ns;
        if (ns !== 0) slot.has_net_sales = true;
      }
      byDate[row.date] = slot;
    }
  }
  return Object.values(byDate).sort((a, b) => a.date.localeCompare(b.date));
}

// Sum sales_summary.daily (Toast-imported Net Sales baseline) across RCs.
// deno-lint-ignore no-explicit-any
function aggregateSalesSummaryDaily(payload: any, dateRange: Set<string>): { date: string; net_sales: number }[] {
  const ss = payload?.sales_summary;
  if (!ss || typeof ss !== "object") return [];
  const byDate: Record<string, { date: string; net_sales: number }> = {};
  for (const rcKey of Object.keys(ss)) {
    const rcDaily = ss[rcKey]?.daily;
    if (!Array.isArray(rcDaily)) continue;
    for (const row of rcDaily) {
      if (!row || typeof row.date !== "string") continue;
      if (!dateRange.has(row.date)) continue;
      const slot = byDate[row.date] ?? { date: row.date, net_sales: 0 };
      const ns = Number(row.net_sales);
      if (Number.isFinite(ns)) slot.net_sales += ns;
      byDate[row.date] = slot;
    }
  }
  return Object.values(byDate).sort((a, b) => a.date.localeCompare(b.date));
}

export async function runCrossSourceReconciler(): Promise<ReconcilerResult> {
  const audits: AuditDecision[] = [];
  const alerts: ReconcilerAlert[] = [];
  const ts = new Date().toISOString();
  const dateRange = new Set(lastNDays(LOOKBACK_DAYS));

  for (const outlet of OUTLETS) {
    // deno-lint-ignore no-explicit-any
    let payload: Record<string, any>;
    try {
      const r = await fetch(`${PAGES_BASE}/${outlet}.json`, { cache: "no-cache" });
      if (!r.ok) {
        audits.push({
          ts, agent: "cross_source_reconciler", source: outlet,
          decision: "fetch_failed",
          details: { http_status: r.status },
          action_taken: "skipped this tick",
        });
        continue;
      }
      payload = await r.json();
    } catch (e) {
      audits.push({
        ts, agent: "cross_source_reconciler", source: outlet,
        decision: "fetch_error",
        details: { error: String(e) },
        action_taken: "skipped this tick",
      });
      continue;
    }

    const odDaily = aggregateOrderDetailsDaily(payload, dateRange);
    const ssDaily = aggregateSalesSummaryDaily(payload, dateRange);

    // ====================================================================
    // CHECK 1: internal ratio (amount / net_sales)
    // ====================================================================
    const odWithNS = odDaily.filter((r) => r.has_net_sales);
    if (odWithNS.length < MIN_NET_SALES_DAYS) {
      // Not enough post-fix coverage yet. Note in audit but don't alert.
      audits.push({
        ts, agent: "cross_source_reconciler", source: outlet,
        decision: "internal_check_skipped_insufficient_coverage",
        details: {
          days_with_net_sales: odWithNS.length,
          min_required: MIN_NET_SALES_DAYS,
          lookback_days: LOOKBACK_DAYS,
        },
        action_taken: "deferred until backfill populates net_sales on more rows",
      });
    } else {
      const totalAmt = odWithNS.reduce((s, r) => s + r.amount, 0);
      const totalNS = odWithNS.reduce((s, r) => s + r.net_sales, 0);
      const ratio = totalNS > 0 ? totalAmt / totalNS : null;
      if (ratio == null || !Number.isFinite(ratio)) {
        audits.push({
          ts, agent: "cross_source_reconciler", source: outlet,
          decision: "internal_check_zero_net_sales",
          details: { totalAmt, totalNS, days: odWithNS.length },
          action_taken: "zero or non-finite Net Sales — manual check warranted",
        });
      } else if (ratio < RATIO_LOWER || ratio > RATIO_UPPER) {
        const text = `amount/net_sales ratio = ${ratio.toFixed(2)} ` +
          `(amount $${totalAmt.toFixed(0)} / net_sales $${totalNS.toFixed(0)} ` +
          `over ${odWithNS.length} days). Expected band [${RATIO_LOWER}, ${RATIO_UPPER}]. ` +
          `Likely ETL regression — net_sales formula or check.amount semantics changed.`;
        alerts.push({ outlet, kind: "internal_ratio_breach", text });
        audits.push({
          ts, agent: "cross_source_reconciler", source: outlet,
          decision: "internal_ratio_breach",
          details: { ratio, totalAmt, totalNS, days: odWithNS.length,
                     band: [RATIO_LOWER, RATIO_UPPER] },
          action_taken: "queued for Slack alert",
        });
      } else {
        audits.push({
          ts, agent: "cross_source_reconciler", source: outlet,
          decision: "internal_ratio_ok",
          details: { ratio, totalAmt, totalNS, days: odWithNS.length },
          action_taken: "no alert",
        });
      }
    }

    // ====================================================================
    // CHECK 2: external — order_details.net_sales vs sales_summary.net_sales
    // ====================================================================
    if (ssDaily.length === 0) {
      // No sales_summary baseline. Skip silently — once the auto-pull
      // lands this'll start running.
    } else {
      // Find dates where BOTH sources have data and order_details has net_sales populated.
      const odMap: Record<string, DailyTotals> = {};
      for (const r of odDaily) odMap[r.date] = r;
      const overlap: { date: string; od: number; ss: number }[] = [];
      for (const ss of ssDaily) {
        const od = odMap[ss.date];
        if (!od || !od.has_net_sales) continue;
        overlap.push({ date: ss.date, od: od.net_sales, ss: ss.net_sales });
      }

      if (overlap.length < MIN_NET_SALES_DAYS) {
        audits.push({
          ts, agent: "cross_source_reconciler", source: outlet,
          decision: "external_check_skipped_insufficient_overlap",
          details: { overlap_days: overlap.length, min_required: MIN_NET_SALES_DAYS },
          action_taken: "deferred until both sources cover >=7 of last 30 days",
        });

        // Stale-baseline check: if sales_summary's most recent date is
        // > 7 days behind today, flag it. Once the auto-pull is live this
        // shouldn't happen — alerts here motivate fixing the upstream.
        const ssLatest = ssDaily[ssDaily.length - 1]?.date;
        if (ssLatest) {
          const ageDays = Math.floor(
            (Date.now() - new Date(ssLatest + "T12:00:00Z").getTime()) / 86400000,
          );
          if (ageDays > 7) {
            const text = `sales_summary baseline is ${ageDays} days stale ` +
              `(latest ${ssLatest}). Refresh the Toast Sales Summary CSV import ` +
              `or accelerate the auto-pull rollout.`;
            alerts.push({ outlet, kind: "stale_baseline", text });
            audits.push({
              ts, agent: "cross_source_reconciler", source: outlet,
              decision: "stale_baseline",
              details: { ss_latest: ssLatest, age_days: ageDays },
              action_taken: "queued for Slack alert",
            });
          }
        }
      } else {
        const totalOd = overlap.reduce((s, r) => s + r.od, 0);
        const totalSs = overlap.reduce((s, r) => s + r.ss, 0);
        if (totalSs <= 0) {
          audits.push({
            ts, agent: "cross_source_reconciler", source: outlet,
            decision: "external_check_zero_baseline",
            details: { totalOd, totalSs, days: overlap.length },
            action_taken: "sales_summary contributed zero — likely import regression",
          });
        } else {
          const drift = (totalOd - totalSs) / totalSs;
          if (Math.abs(drift) > CROSS_SOURCE_DRIFT) {
            const text = `cross-source drift = ${(drift * 100).toFixed(1)}% ` +
              `(order_details $${totalOd.toFixed(0)} vs sales_summary $${totalSs.toFixed(0)} ` +
              `over ${overlap.length} overlapping days). Threshold ±${(CROSS_SOURCE_DRIFT * 100).toFixed(0)}%. ` +
              `One of the two Net Sales feeds is stale or misaligned.`;
            alerts.push({ outlet, kind: "external_source_drift", text });
            audits.push({
              ts, agent: "cross_source_reconciler", source: outlet,
              decision: "external_source_drift",
              details: { drift, totalOd, totalSs, days: overlap.length,
                         threshold: CROSS_SOURCE_DRIFT },
              action_taken: "queued for Slack alert",
            });
          } else {
            audits.push({
              ts, agent: "cross_source_reconciler", source: outlet,
              decision: "external_drift_ok",
              details: { drift, totalOd, totalSs, days: overlap.length },
              action_taken: "no alert",
            });
          }
        }
      }
    }
  }

  return { audits, alerts };
}

File: toast-etl/cross_source_reconciler.py
"""Cross-source reconciler — Python port for GH Actions.

Twin of supabase/functions/agent-worker/agents/cross_source_reconciler.ts
that runs via GH Actions instead of the Supabase Edge Function. Lets the
reconciliation layer work even when the Edge Function isn't deployed
(Phase A.1 deploy is gated on Slack admin approval; this script is not).

Per outlet, last 30 days:

  1. INTERNAL: order_details.amount / order_details.net_sales ratio.
     Sanity bounds [0.90, 1.50]. Breach → ETL regression flag.

  2. EXTERNAL: order_details.net_sales vs sales_summary.net_sales on
     overlapping dates. Drift threshold ±5%. Breach → source-feed
     mismatch (one of the two is stale or misaligned).

  3. STALE: if sales_summary.daily latest date > 7 days behind today,
     flag the upstream feed as stale.

Outputs a reconciliation report to data/_validation/_reconciliation.json
that the dashboard reads + the morning briefing email pulls from. NO
direct Slack posting (skipped per Ross's note: pending admin approval).

When the Edge Function deploys, this script becomes redundant with the
TS agent — they produce identical audits. Two parallel implementations
is fine for now (defense in depth) and the Edge Function can be deactivated
later if desired.
"""
from __future__ import annotations

import json
import sys
from datetime import date, datetime, timedelta, timezone
from pathlib import Path
from typing import Any

DATA_DIR = Path(__file__).resolve().parent.parent / "data"
VALIDATION_DIR = DATA_DIR / "_validation"
OUT_PATH = VALIDATION_DIR / "_reconciliation.json"

OUTLETS = ["lsbr", "mulherins", "kampers", "lowland", "vessel", "anthology",
           "rosemary_rose", "hiroki_det", "hiroki_phl", "little_wing", "quoin"]

# Sanity bounds for amount/net_sales ratio. Healthy LSBR sits ~1.02
# post-PR-#100 (UI-aligned formula); the original bug had it at 1.28.
# A breach above 1.50 means check.amount started including something
# new the formula doesn't subtract. Below 0.90 means net_sales somehow
# exceeds amount — schema regression or sign error.
RATIO_LOWER = 0.90
RATIO_UPPER = 1.50

# Cross-source drift threshold: 5% sustained delta across the 30-day
# lookback between two independently-computed Net Sales values.
CROSS_SOURCE_DRIFT = 0.05

# Minimum coverage before checks are meaningful.
MIN_DAYS = 7
LOOKBACK_DAYS = 30


def _aggregate_order_details(payload: dict, dates: set[str]) -> list[dict]:
    """Sum order_details.daily across all RC keys per date."""
    od = payload.get("order_details") or {}
    by_date: dict[str, dict] = {}
    for _rc, b in od.items():
        for r in b.get("daily", []) or []:
            d = r.get("date") or ""
            if d not in dates:
                continue
            slot = by_date.setdefault(d, {"date": d, "amount": 0.0, "net_sales": 0.0,
                                           "has_net_sales": False})
            slot["amount"]    += float(r.get("amount") or 0)
            ns = r.get("net_sales")
            if ns is not None:
                slot["net_sales"] += float(ns)
                if float(ns) != 0:
                    slot["has_net_sales"] = True
    return sorted(by_date.values(), key=lambda r: r["date"])


def _aggregate_sales_summary(payload: dict, dates: set[str]) -> list[dict]:
    """Sum sales_summary.daily across all RC keys per date."""
    ss = payload.get("sales_summary") or {}
    by_date: dict[str, dict] = {}
    for _rc, b in ss.items():
        for r in b.get("daily", []) or []:
            d = r.get("date") or ""
            if d not in dates:
                continue
            slot = by_date.setdefault(d, {"date": d, "net_sales": 0.0})
            slot["net_sales"] += float(r.get("net_sales") or 0)
    return sorted(by_date.values(), key=lambda r: r["date"])


def reconcile_outlet(outlet: str) -> dict:
    """Run all three checks against one outlet, return a structured result."""
    payload_path = DATA_DIR / f"{outlet}.json"
    if not payload_path.exists():
        return {"outlet": outlet, "checks": [{"kind": "missing_payload",
                "status": "skip", "detail": f"data/{outlet}.json not found"}]}
    payload = json.loads(payload_path.read_text(encoding="utf-8"))

    today = date.today()
    dates = {(today - timedelta(days=i)).isoformat() for i in range(1, LOOKBACK_DAYS + 1)}
    od = _aggregate_order_details(payload, dates)
    ss = _aggregate_sales_summary(payload, dates)

    checks: list[dict] = []

    # --- Check 1: internal amount/net_sales ratio ---
    od_with_ns = [r for r in od if r["has_net_sales"]]
    if len(od_with_ns) < MIN_DAYS:
        checks.append({"kind": "internal_ratio", "status": "skip",
                       "detail": f"only {len(od_with_ns)}/{LOOKBACK_DAYS} days with "
                                 f"net_sales populated (need ≥{MIN_DAYS}); "
                                 f"deferred until backfill catches up"})
    else:
        amt = sum(r["amount"] for r in od_with_ns)
        ns  = sum(r["net_sales"] for r in od_with_ns)
        if ns > 0:
            ratio = amt / ns
            band = (RATIO_LOWER, RATIO_UPPER)
            in_band = RATIO_LOWER <= ratio <= RATIO_UPPER
            checks.append({
                "kind": "internal_ratio",
                "status": "ok" if in_band else "alert",
                "ratio": round(ratio, 4),
                "amount_30d": round(amt, 2),
                "net_sales_30d": round(ns, 2),
                "days_compared": len(od_with_ns),
                "band": band,
                "detail": (f"amount/net_sales = {ratio:.3f} "
                           f"(${amt:,.0f}/${ns:,.0f} over {len(od_with_ns)}d)"
                           + ("" if in_band
                              else " — OUTSIDE [0.90, 1.50] sanity band")),
            })
        else:
            checks.append({"kind": "internal_ratio", "status": "alert",
                           "detail": f"net_sales sum is {ns:.2f} — "
                                     "either zero revenue or schema regression"})

    # --- Check 2: external order_details vs sales_summary ---
    if not ss:
        checks.append({"kind": "external_drift", "status": "skip",
                       "detail": "no sales_summary baseline available "
                                 "(awaiting Toast Sales Summary auto-pull)"})
    else:
        od_map = {r["date"]: r for r in od}
        overlap = []
        for r in ss:
            o = od_map.get(r["date"])
            if not o or not o["has_net_sales"]:
                continue
            overlap.append({"date": r["date"], "od": o["net_sales"], "ss": r["net_sales"]})
        if len(overlap) < MIN_DAYS:
            checks.append({"kind": "external_drift", "status": "skip",
                           "detail": f"only {len(overlap)} overlapping days with both "
                                     f"sources (need ≥{MIN_DAYS})"})
        else:
            tot_od = sum(r["od"] for r in overlap)
            tot_ss = sum(r["ss"] for r in overlap)
            if tot_ss > 0:
                drift = (tot_od - tot_ss) / tot_ss
                ok = abs(drift) <= CROSS_SOURCE_DRIFT
                checks.append({
                    "kind": "external_drift",
                    "status": "ok" if ok else "alert",
                    "drift_pct": round(drift * 100, 4),
                    "od_total": round(tot_od, 2),
                    "ss_total": round(tot_ss, 2),
                    "overlap_days": len(overlap),
                    "threshold_pct": CROSS_SOURCE_DRIFT * 100,
                    "detail": (f"order_details ${tot_od:,.0f} vs "
                               f"sales_summary ${tot_ss:,.0f} = {drift*100:+.2f}% drift"
                               + ("" if ok
                                  else f" — exceeds ±{CROSS_SOURCE_DRIFT*100:.0f}%")),
                })
            else:
                checks.append({"kind": "external_drift", "status": "alert",
                               "detail": "sales_summary baseline is zero — import regression"})

        # --- Check 3: stale baseline ---
        ss_latest = max((r["date"] for r in ss), default=None)
        if ss_latest:
            age = (today - date.fromisoformat(ss_latest)).days
            ok = age <= 7
            checks.append({
                "kind": "stale_baseline",
                "status": "ok" if ok else "alert",
                "ss_latest_date": ss_latest,
                "age_days": age,
                "detail": (f"sales_summary latest = {ss_latest} ({age}d ago)"
                           + ("" if ok else f" — exceeds 7d freshness threshold")),
            })

    return {"outlet": outlet, "checks": checks}


def main() -> int:
    VALIDATION_DIR.mkdir(parents=True, exist_ok=True)
    ran_at = datetime.now(timezone.utc).isoformat(timespec="seconds")
    report = {
        "ran_at": ran_at,
        "lookback_days": LOOKBACK_DAYS,
        "sanity_bounds": {"ratio_lower": RATIO_LOWER, "ratio_upper": RATIO_UPPER,
                          "drift_threshold": CROSS_SOURCE_DRIFT},
        "outlets": [],
    }
    n_alerts = 0
    n_ok = 0
    for outlet in OUTLETS:
        result = reconcile_outlet(outlet)
        for c in result["checks"]:
            if c["status"] == "alert":
                n_alerts += 1
            elif c["status"] == "ok":
                n_ok += 1
        report["outlets"].append(result)
    report["summary"] = {"checks_ok": n_ok, "checks_alert": n_alerts,
                         "outlets_reconciled": len(OUTLETS)}

    OUT_PATH.write_text(json.dumps(report, indent=2), encoding="utf-8")
    sys.stdout.write(f"wrote {OUT_PATH} — {n_ok} OK, {n_alerts} alerts across "
                     f"{len(OUTLETS)} outlets\n")

    # Print per-outlet summary to stdout for the workflow log
    for r in report["outlets"]:
        outlet = r["outlet"]
        for c in r["checks"]:
            status_glyph = {"ok": "OK", "alert": "ALERT", "skip": "skip"}[c["status"]]
            sys.stdout.write(f"  [{outlet:<14}] {c['kind']:<18} {status_glyph:<5} "
                             f"{c.get('detail', '')}\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
